File: csharp/src/DataStructures/DoublyLinkedList.cs
namespace DSA.DataStructures;

/// <summary>
/// Doubly linked list implementation.
/// Time: O(1) for front/back ops, O(n) for search
/// Space: O(n)
/// </summary>
public class DoublyLinkedList<T>
{
    private class Node
    {
        public T Value { get; set; }
        public Node? Prev { get; set; }
        public Node? Next { get; set; }

        public Node(T value)
        {
            Value = value;
        }
    }

    private Node? _head;
    private Node? _tail;
    private int _size;

    public int Count => _size;
    public bool IsEmpty => _size == 0;

    public void PushFront(T value)
    {
        var node = new Node(value) { Next = _head };
        if (_head != null)
            _head.Prev = node;
        else
            _tail = node;
        _head = node;
        _size++;
    }

    public void PushBack(T value)
    {
        var node = new Node(value) { Prev = _tail };
        if (_tail != null)
            _tail.Next = node;
        else
            _head = node;
        _tail = node;
        _size++;
    }

    public T? PopFront()
    {
        if (_head == null)
            return default;

        var value = _head.Value;
        _head = _head.Next;
        if (_head != null)
            _head.Prev = null;
        else
            _tail = null;
        _size--;
        return value;
    }

    public T? PopBack()
    {
        if (_tail == null)
            return default;

        var value = _tail.Value;
        _tail = _tail.Prev;
        if (_tail != null)
            _tail.Next = null;
        else
            _head = null;
        _size--;
        return value;
    }

    public T? PeekFront() => _head == null ? default : _head.Value;
    public T? PeekBack() => _tail == null ? default : _tail.Value;

    public bool Contains(T value)
    {
        var current = _head;
        while (current != null)
        {
            if (EqualityComparer<T>.Default.Equals(current.Value, value))
                return true;
            current = current.Next;
        }
        return false;
    }

    public bool Remove(T value)
    {
        var current = _head;
        while (current != null)
        {
            if (EqualityComparer<T>.Default.Equals(current.Value, value))
            {
                if (current.Prev != null)
                    current.Prev.Next = current.Next;
                else
                    _head = current.Next;

                if (current.Next != null)
                    current.Next.Prev = current.Prev;
                else
                    _tail = current.Prev;

                _size--;
                return true;
            }
            current = current.Next;
        }
        return false;
    }

    public List<T> ToList()
    {
        var result = new List<T>();
        var current = _head;
        while (current != null)
        {
            result.Add(current.Value);
            current = current.Next;
        }
        return result;
    }

    public void Clear()
    {
        _head = null;
        _tail = null;
        _size = 0;
    }
}

File: php/src/DataStructures/DoublyLinkedList.php
<?php

declare(strict_types=1);

namespace DSA\DataStructures;

/**
 * Doubly linked list implementation.
 * Time: O(1) for front/back ops, O(n) for search
 * Space: O(n)
 *
 * @template T
 */
class DoublyLinkedList
{
    private ?DoublyNode $head = null;
    private ?DoublyNode $tail = null;
    private int $size = 0;

    public function pushFront(mixed $value): void
    {
        $node = new DoublyNode($value, next: $this->head);
        if ($this->head !== null) {
            $this->head->prev = $node;
        } else {
            $this->tail = $node;
        }
        $this->head = $node;
        $this->size++;
    }

    public function pushBack(mixed $value): void
    {
        $node = new DoublyNode($value, prev: $this->tail);
        if ($this->tail !== null) {
            $this->tail->next = $node;
        } else {
            $this->head = $node;
        }
        $this->tail = $node;
        $this->size++;
    }

    public function popFront(): mixed
    {
        if ($this->head === null) {
            return null;
        }
        $value = $this->head->value;
        $this->head = $this->head->next;
        if ($this->head !== null) {
            $this->head->prev = null;
        } else {
            $this->tail = null;
        }
        $this->size--;
        return $value;
    }

    public function popBack(): mixed
    {
        if ($this->tail === null) {
            return null;
        }
        $value = $this->tail->value;
        $this->tail = $this->tail->prev;
        if ($this->tail !== null) {
            $this->tail->next = null;
        } else {
            $this->head = null;
        }
        $this->size--;
        return $value;
    }

    public function peekFront(): mixed
    {
        return $this->head?->value;
    }

    public function peekBack(): mixed
    {
        return $this->tail?->value;
    }

    public function contains(mixed $value): bool
    {
        $current = $this->head;
        while ($current !== null) {
            if ($current->value === $value) {
                return true;
            }
            $current = $current->next;
        }
        return false;
    }

    public function remove(mixed $value): bool
    {
        $current = $this->head;
        while ($current !== null) {
            if ($current->value === $value) {
                if ($current->prev !== null) {
                    $current->prev->next = $current->next;
                } else {
                    $this->head = $current->next;
                }
                if ($current->next !== null) {
                    $current->next->prev = $current->prev;
                } else {
                    $this->tail = $current->prev;
                }
                $this->size--;
                return true;
            }
            $current = $current->next;
        }
        return false;
    }

    public function isEmpty(): bool
    {
        return $this->size === 0;
    }

    public function size(): int
    {
        return $this->size;
    }

    public function toArray(): array
    {
        $result = [];
        $current = $this->head;
        while ($current !== null) {
            $result[] = $current->value;
            $current = $current->next;
        }
        return $result;
    }

    public function clear(): void
    {
        $this->head = null;
        $this->tail = null;
        $this->size = 0;
    }
}

/**
 * @internal
 */
class DoublyNode
{
    public function __construct(
        public mixed $value,
        public ?DoublyNode $prev = null,
        public ?DoublyNode $next = null
    ) {}
}
